src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs

```diff
@@ -96,7 +98,10 @@ public virtual int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsH
                     testSources = testRunCriteria.Tests.GroupBy(tc => tc.Source).Select(g => g.Key);
                 }
 
-                this.isCommunicationEstablished = this.SetupChannel(testSources, this.cancellationTokenSource.Token);
+                lock (locObject)
```

Comment:
Consider a scenario where Channel is getting setup (testhost has been launched but setup channel didn't return back ( First thread is at this point [code](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs#L130)) ) and in between Cancelation request come. Since at this point channel is not setup completely cancel function will not call cancel of test host ( Second thread is at this point [code](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyExecutionManager.cs#L166))

To deal this situation we need lock here